src/txrequest.h

```diff
@@ -0,0 +1,342 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_TXREQUEST_H
+#define BITCOIN_TXREQUEST_H
+
+#include <crypto/siphash.h>
+#include <uint256.h>
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/mem_fun.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+
+#include <chrono>
+#include <unordered_map>
+#include <vector>
+
+#include <stdint.h>
+
+
+/** A functor with embedded salt that computes priority of a txid/peer combination.
+ *
+ * The priority is used to order candidates for a given txid; lower priority values are selected first. It has the
+ * following properties:
+ * - Priorities for outbound peers are always lower than inbound peers, making them always preferred.
+ * - Within the set of outbound peers and within the set of inbound peers:
+ *   - The very first announcement for a given txid is always lowest.
+ *   - Announcements after the first one are ordered uniformly randomly (using a salted hash of txid/peer).
+ */
+class PriorityComputer {
+    const uint64_t m_k0, m_k1;
+public:
+    explicit PriorityComputer(bool deterministic);
+    uint64_t operator()(const uint256& txid, uint64_t peer, bool inbound, bool first) const
+    {
+        uint64_t low_bits = 0;
+        if (!first) {
+            low_bits = CSipHasher(m_k0, m_k1).Write(txid.begin(), txid.size()).Write(peer).Finalize() >> 1;
```

Comment:
272a615fad6f93e1b7f9b6245ce9090d07fb4a51

There is `SipHashUint256Extra(uint64_t, uint64_t, uint256&, uint32_t)` in the `crypto/siphash` module.
Is it worth to add the reusable `SipHashUint256Extra64(uint64_t, uint64_t, uint256&, uint64_t)` there?

Reply:
I think this may make sense as a follow-up, but probably deserves benchmarks etc.

Reply:
Somewhat related ;) https://github.com/bitcoin/bitcoin/pull/18014